src/VisualStudio/Xaml/Impl/Implementation/XamlProjectService.cs

```diff
@@ -83,54 +84,70 @@ public void TrackOpenDocument(string filePath)
+                if (hierarchy == null)
                 {
                     return;
                 }
 
-                var projectInfo = new VisualStudioProjectCreationInfo
+                if (!_xamlProjects.TryGetValue(hierarchy, out var project))
                 {
-                    Hierarchy = hierarchy,
-                    FilePath = hierarchy.TryGetProjectFilePath(),
-                    ProjectGuid = projectGuid
-                };
-
-                project = _visualStudioProjectFactory.CreateAndAddToWorkspace(name, StringConstants.XamlLanguageName, projectInfo);
-                _xamlProjects.Add(hierarchy, project);
-            }
-
-            if (!project.ContainsSourceFile(filePath))
-            {
-                project.AddSourceFile(filePath);
+                    if (!hierarchy.TryGetName(out var name))
+                    {
+                        return;
+                    }
+
+                    if (!hierarchy.TryGetGuidProperty(__VSHPROPID.VSHPROPID_ProjectIDGuid, out var projectGuid))
+                    {
+                        return;
+                    }
+
+                    var projectInfo = new VisualStudioProjectCreationInfo
+                    {
+                        Hierarchy = hierarchy,
+                        FilePath = hierarchy.TryGetProjectFilePath(),
+                        ProjectGuid = projectGuid
+                    };
+
+                    project = _visualStudioProjectFactory.CreateAndAddToWorkspace(name, StringConstants.XamlLanguageName, projectInfo);
+                    _xamlProjects.Add(hierarchy, project);
+                }
 
-                var documentId = _workspace.CurrentSolution.GetDocumentIdsWithFilePath(filePath).First(d => d.ProjectId == project.Id);
-                var document = _workspace.CurrentSolution.GetDocument(documentId)!;
-                var hasText = document.TryGetText(out var text);
-                if (!hasText || text?.Container.TryGetTextBuffer() == null)
+                if (!project.ContainsSourceFile(filePath))
                 {
-                    var docInfo = _rdt.GetDocumentInfo(docCookie);
-                    var textBuffer = _editorAdaptersFactory.GetDocumentBuffer(docInfo.DocData as IVsTextBuffer);
-                    var textContainer = textBuffer.AsTextContainer();
-                    _workspace.OnDocumentTextChanged(documentId, textContainer.CurrentText, PreservationMode.PreserveIdentity);
+                    project.AddSourceFile(filePath);
+
+                    var documentId = _workspace.CurrentSolution.GetDocumentIdsWithFilePath(filePath).Single(d => d.ProjectId == project.Id);
+                    var document = _workspace.CurrentSolution.GetDocument(documentId)!;
+                    var hasText = document.TryGetText(out var text);
+                    if (!hasText || text?.Container.TryGetTextBuffer() == null)
+                    {
```

Comment:
I'm confused by why this condition is here -- I would have expected that since you're adding the document by file path we always would be hooking up to the RDT text buffer, so you shouldn't have to write this at all. Is that sometimes not working?

Reply:
Or put another way: why is this done here but not in the logic handling a rename?

Reply:
Correct, it isn't hooked up yet to the actual VS buffer the first time we ask for it after a document is opened.

Reply:
Rename handling is more about updating the documents we know about in XamlProjects. TrackOpenDocument handles the case where a XAML document has been opened and there we need to ensure that it is hooked up to the real VS buffer at this time.

Reply:
But why is the logic conditioned versus just always doing it? Does it imply that sometimes this already _is_ connected to an open file for some reason? I guess doing the OnDocumentTextChanged that this is calling is really not a supported action, since the VisualStudioProject owns the contents and could undo the work being done.

Reply:
This isn't needed when you reopen a document. Should I call a different API to hook up the buffer?

Reply:
FWIW, this seems to work just fine, and it really just associates it with the ITextBuffer that it will eventually get anyway.

Reply:
@jasonmalinowski Can this be merged now?